File: src/dotnet/UI.Blazor/Controls/VirtualList/VirtualListData.cs
namespace ActualChat.UI.Blazor.Controls;

public class VirtualListData<TItem>
    where TItem : IVirtualListItem
{
#pragma warning disable MA0018
    public static VirtualListData<TItem> None { get; } = new(VirtualListDataQuery.None, Array.Empty<TItem>());
#pragma warning restore MA0018

    public VirtualListDataQuery Query { get; }
    public IReadOnlyCollection<TItem> Items { get; }
    public bool HasVeryFirstItem { get; init; }
    public bool HasVeryLastItem { get; init; }
    public bool HasAllItems => HasVeryFirstItem && HasVeryLastItem;
    public Symbol ScrollToKey { get; init; }

    public VirtualListData(VirtualListDataQuery query, IReadOnlyCollection<TItem> items)
    {
        Query = query;
        Items = items;
    }
}

public static class VirtualListData
{
    public static VirtualListData<TItem> New<TItem>(
        VirtualListDataQuery query,
        IEnumerable<TItem> items,
        bool hasVeryFirstItem = false,
        bool hasVeryLastItem = false,
        Symbol scrollToKey = default)
        where TItem : IVirtualListItem
    {
        var readOnlyItems = items as IReadOnlyCollection<TItem> ?? items.ToList();
        return new(query, readOnlyItems) {
            HasVeryFirstItem = hasVeryFirstItem,
            HasVeryLastItem = hasVeryLastItem,
            ScrollToKey = scrollToKey,
        };
    }
}

File: src/dotnet/UI.Blazor/Components/VirtualList/ts/virtual-list-data.ts
import { VirtualListItem } from './virtual-list-item';
import { VirtualListDataQuery } from './virtual-list-data-query';

export class VirtualListData<TItem extends VirtualListItem>
{
    public static None: VirtualListData<any> = new VirtualListData(VirtualListDataQuery.None, []);

    public readonly Query: VirtualListDataQuery;
    public readonly Items: TItem[];

    public HasVeryFirstItem: boolean = false;
    public HasVeryLastItem: boolean = false
    public ScrollToKey?: string = null;

    constructor(query: VirtualListDataQuery, items: TItem[])
    {
        this.Query = query;
        this.Items = items;
    }

    public get HasAllItems(): boolean {
        return this.HasVeryFirstItem && this.HasVeryLastItem;
    }
}
